tests/fuzz/fuzz-bolt11.c

```diff
@@ -0,0 +1,138 @@
+	       "h6lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq"
+	       "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqxnht6w");
+	size = strlen((char *)fuzz_data);
+
+	clean_tmpctx();
+	return size > max_size ? 0 : size;
+}
+
+// We use a custom mutator to produce an input corpus that consists entirely of
+// correctly encoded bech32 strings. This enables us to efficiently fuzz the
+// bolt11 decoding logic without the fuzzer getting stuck on fuzzing the bech32
+// decoding/encoding logic.
+//
+// This custom mutator does the following things:
+//   1. Attempt to bech32 decode the given input (returns the encoded dummy
+//      invoice on failure).
+//   2. Mutate either the human readable or data part of the invoice using
+//      libFuzzer's default mutator `LLVMFuzzerMutate`.
+//   3. Attempt to bech32 encode the mutated hrp and data (returns the endcoded
+//      dummy on failure).
+//   4. Write the encoded result to `fuzz_data` if its size does not exceed
+//      `max_size`, otherwise return the encoded dummy invoice.
+size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, size_t max_size,
+			       unsigned int seed)
+{
+	if (size < 10)
+		return initial_input(fuzz_data, size, max_size);
+
+	// Interpret fuzz input as string (ensure it's null terminated).
+	char *input = to_string(tmpctx, fuzz_data, size);
+
+	// Attempt to bech32 decode the input.
+	char *hrp = tal_arr(tmpctx, char, strlen(input) - 6);
+	u5 *data = tal_arr(tmpctx, u5, strlen(input) - 8);
+	size_t datalen = 0;
+	if (bech32_decode(hrp, data, &datalen, input, (size_t)-1) !=
+	    BECH32_ENCODING_BECH32) {
+		// Decoding failed, this should only happen when starting from
+		// an empty corpus.
+		return initial_input(fuzz_data, size, max_size);
+	}
+
+	// Mutate either the hrp or data. Given the same seed, the same
+	// mutation is performed.
+	srand(seed);
+	switch (rand() % 2) {
+	case 0: // Mutate hrp and ensure it's still null terminated.
+		hrp[LLVMFuzzerMutate((uint8_t *)hrp, strlen(hrp),
+				     strlen(input) - 6) -
+		    1] = '\0';
+		break;
+	case 1: // Mutate data and re-assign datalen.
+		datalen = LLVMFuzzerMutate((uint8_t *)data, datalen,
+					   strlen(input) - 8);
+		break;
+	}
+
+	// Encode the mutated input.
+	char *output = tal_arr(tmpctx, char, strlen(hrp) + datalen + 8);
+	if (!bech32_encode(output, hrp, data, datalen, (size_t)-1,
+			   BECH32_ENCODING_BECH32)) {
+		return initial_input(fuzz_data, size, max_size);
+	}
+
+	// Strip (repeated) "lightning:" prefixes
+	while (1) {
+		char *stripped =
+		    (char *)to_canonical_invstr(tmpctx, (const char *)output);
+		if (memcmp(stripped, output, strlen(stripped)) == 0) {
+			// No "lightning:" prefix left
+			break;
+		}
+
+		output = stripped;
+	}
+
+	// Write the result into `fuzz_data`.
+	size_t output_len = strlen(output);
+	if (output_len >= max_size)
+		return initial_input(fuzz_data, size, max_size);
+
+	strcpy((char *)fuzz_data, output);
```

Comment:
We actually don't need to copy the null byte, since returning `output_len` causes it to be ignored anyway.
```suggestion
	if (output_len > max_size)
		return initial_input(fuzz_data, size, max_size);

	memcpy((char *)fuzz_data, output, output_len);
```